composer/optim/scheduler.py

```diff
@@ -1,226 +1,265 @@
+    current_time = state.timer.get(T_max.unit)
+    frac_of_total = (current_time / T_max).value
+
+    return _cosine_anneal(x=frac_of_total, min_y=min_factor)
+
+
+def cosine_annealing_warm_restarts_scheduler(state: State,
+                                             *,
+                                             ssr: float = 1.0,
+                                             T_0: Union[str, Time],
+                                             T_mult: float = 1.0,
+                                             min_factor: float = 0.0):
+    T_0 = _convert_time(T_0, state, ssr=ssr)
+    current_interval_len = T_0
+    current_interval_end = T_0
+    while current_interval_end <= state.timer.get(current_interval_end.unit):
+        if current_interval_len.value == 0:
+            raise ValueError('Interval between restarts for cosine annealing/warm restarts scheduler has decayed to 0.')
+
+        current_interval_len = Time(value=int(T_mult * current_interval_len.value), unit=current_interval_len.unit)
+        current_interval_end += current_interval_len
+
+    current_interval_start = current_interval_end - current_interval_len
+    frac_of_current_interval = ((state.timer.get(T_0.unit) - current_interval_start) / current_interval_len).value
+
+    return _cosine_anneal(x=frac_of_current_interval, min_y=min_factor)
+
+
+def polynomial_scheduler(state: State,
+                         *,
+                         ssr: float = 1.0,
+                         T_max: Union[str, Time] = '1dur',
+                         power: float,
+                         min_factor: float = 0.0):
+    T_max = _convert_time(T_max, state, ssr=ssr)
+    current_time = state.timer.get(T_max.unit)
+    frac_of_total = (current_time / T_max).value
+
+    coeff = (1 - frac_of_total)**power
+    current_factor = min_factor + coeff * (1.0 - min_factor)
+    return current_factor
+
+
+def multi_step_with_warmup_scheduler(state: State,
+                                     *,
+                                     ssr: float = 1.0,
+                                     warmup_time: Union[str, Time],
+                                     milestones: List[Union[str, Time]],
+                                     gamma: float = 0.1) -> float:
+    warmup_time = _convert_time(warmup_time, state)
+    if state.timer < warmup_time:
+        return linear_scheduler(state, start_factor=0.0, end_factor=1.0, total_time=warmup_time)
+
+    return multi_step_scheduler(state, ssr=ssr, milestones=milestones, gamma=gamma)
+
+
+def linear_with_warmup_scheduler(state: State,
+                                 *,
+                                 ssr: float = 1.0,
+                                 warmup_time: Union[str, Time],
+                                 start_factor: float = 1.0,
+                                 end_factor: float = 0.0,
+                                 total_time: Union[str, Time] = '1dur'):
+    # N.B. warmup time is intentionally *not* subject to scale schedule
```

Comment:
N.B. ? 

Should probably just include the detail about warmup in the doctoring for this function (+ all others that deal with warmup like this)

Reply:
[Nota bene](https://en.wikipedia.org/wiki/Nota_bene) - I guess I can find a more normal way to say this. :joy: